ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -806,32 +854,98 @@ toCBORAHeaderOrBoundary pm epochSlots abob =
+
+instance FromCBOR ByronApplyTxError where
+  fromCBOR = do
+    enforceSize "ByronApplyTxError" 2
+    CBOR.decodeWord8 >>= \case
+      0   -> ByronApplyTxError             <$> fromCBOR
+      1   -> ByronApplyDlgError            <$> fromCBOR
+      2   -> ByronApplyUpdateProposalError <$> fromCBOR
+      3   -> ByronApplyUpdateVoteError     <$> fromCBOR
+      tag -> cborError $ DecoderErrorUnknownTag "ByronApplyTxError" tag
+
 instance (ByronGiven, Typeable cfg, ConfigContainsGenesis cfg)
       => ApplyTx (ByronBlockOrEBB cfg) where
   -- | Generalized transactions in Byron
   --
   -- TODO #514: This is still missing the other cases (this shouldn't be a
   -- newtype)
-  data GenTx (ByronBlockOrEBB cfg) = ByronTx
-    { byronTx   :: !(CC.UTxO.ATxAux ByteString)
-    , byronTxId :: CC.UTxO.TxId
-      -- ^ This field is lazy on purpose so that the TxId is computed on
-      -- demand.
-    }
-
-  newtype GenTxId (ByronBlockOrEBB cfg) = ByronTxId { unByronTxId :: CC.UTxO.TxId }
+  data GenTx (ByronBlockOrEBB cfg)
+    = ByronTx
+        CC.UTxO.TxId
+        -- ^ This field is lazy on purpose so that the 'CC.UTxO.TxId' is
+        -- computed on demand.
+        !(CC.UTxO.ATxAux ByteString)
+    | ByronDlg
+        DlgCertId
+        -- ^ This field is lazy on purpose so that the 'DlgCertId' is computed
+        -- on demand.
+        !(CC.Delegation.ACertificate ByteString)
+    | ByronUpdateProposal
+        CC.Update.UpId
+        -- ^ This field is lazy on purpose so that the 'CC.Update.UpId' is
+        -- computed on demand.
+        !(CC.Update.AProposal ByteString)
+    | ByronUpdateVote
+        UpdateVoteId
+        -- ^ This field is lazy on purpose so that the 'UpdateVoteId' is
+        -- computed on demand.
+        !(CC.Update.AVote ByteString)
+
+  data GenTxId (ByronBlockOrEBB cfg)
+    = ByronTxId !CC.UTxO.TxId
+    | ByronDlgId !DlgCertId
+    | ByronUpdateProposalId !CC.Update.UpId
+    | ByronUpdateVoteId !UpdateVoteId
     deriving (Eq, Ord)
 
-  txId = ByronTxId . byronTxId
+  txId (ByronTx txid _)             = ByronTxId txid
+  txId (ByronDlg certHash _)        = ByronDlgId certHash
+  txId (ByronUpdateProposal upid _) = ByronUpdateProposalId upid
+  txId (ByronUpdateVote voteHash _) = ByronUpdateVoteId voteHash
 
-  txSize (ByronTx { byronTx = atxaux }) = fromIntegral txByteSize
+  txSize (ByronTx _ atxaux) = fromIntegral txByteSize
     where
       -- TODO cardano-ledger#576 will provide a function for this
       txByteSize = 1 -- To account for @encodeListLen 2@
                  + (Strict.length . annotation . CC.UTxO.aTaTx      $ atxaux)
                  + (Strict.length . annotation . CC.UTxO.aTaWitness $ atxaux)
 
-  type ApplyTxErr (ByronBlockOrEBB cfg) = CC.UTxO.UTxOValidationError
+  txSize (ByronDlg _ cert) = fromIntegral $
+    Strict.length $ serialize' (void cert)
+
+  txSize (ByronUpdateProposal _ prop) = fromIntegral $
+    Strict.length (recoverBytes prop)
```

Comment:
`AProposal ByteString` is an instance of `Decoded` which allows us to utilize the `recoverBytes` function. Perhaps we should also implement 'Decoded' instances for `AVote ByteString`, `ACertificate ByteString`, and `ATxAux ByteString` to enable a similarly efficient means of calculating the `txSize`.

Note that `AProposal` seems to masquerade as an `Annotated` type in order to maintain a reference to the original `ByteString` from which it was decoded. It performs this masquerade by manually defining a field called `annotation`.

Reply:
Actually, I think they're reworking annotations in `cardano-ledger` so it may be best to leave this for the time being.